threads.go

```diff
@@ -0,0 +1,357 @@
+	logger "github.com/whyrusleeping/go-logging"
+	"google.golang.org/grpc"
+)
+
+var log = logging.Logger("threads")
+
+const (
+	IPEL                     = "ipel"
+	IPELCode                 = 406
+	IPELVersion              = "0.0.1"
+	IPELProtocol protocol.ID = "/" + IPEL + "/" + IPELVersion
+)
+
+var addrProtocol = ma.Protocol{
+	Name:       IPEL,
+	Code:       IPELCode,
+	VCode:      ma.CodeToVarint(IPELCode),
+	Size:       ma.LengthPrefixedVarSize,
+	Transcoder: ma.TranscoderP2P,
+}
+
+func init() {
+	if err := ma.AddProtocol(addrProtocol); err != nil {
+		panic(err)
+	}
+}
+
+type threads struct {
+	host       host.Host
+	blocks     bserv.BlockService
+	service    *service
+	dagService format.DAGService
+	ctx        context.Context
+	cancel     context.CancelFunc
+	tstore.Threadstore
+}
+
+func NewThreads(ctx context.Context, h host.Host, bs bserv.BlockService, ds format.DAGService, ts tstore.Threadstore, debug bool) (tserv.Threadservice, error) {
+	var err error
+	if debug {
+		err = setLogLevels(map[string]logger.Level{
+			"threads":     logger.DEBUG,
+			"threadstore": logger.DEBUG,
+		}, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	t := &threads{
+		host:        h,
+		blocks:      bs,
+		dagService:  ds,
+		ctx:         ctx,
+		cancel:      cancel,
+		Threadstore: ts,
+	}
+	t.service, err = newService(t)
+	if err != nil {
+		return nil, err
+	}
+
+	listener, err := gostream.Listen(h, IPELProtocol)
+	if err != nil {
+		return nil, err
+	}
+	rpc := grpc.NewServer()
+	go rpc.Serve(listener)
+	pb.RegisterThreadsServer(rpc, t.service)
+
+	return t, nil
+}
+
+func (t *threads) Close() (err error) {
+	var errs []error
+	weakClose := func(name string, c interface{}) {
+		if cl, ok := c.(io.Closer); ok {
+			if err = cl.Close(); err != nil {
+				errs = append(errs, fmt.Errorf("%s error: %s", name, err))
+			}
+		}
+	}
+
+	//weakClose("host", t.host) @todo: fix panic on close
+	weakClose("dagservice", t.dagService)
+	weakClose("threadstore", t.Threadstore)
+
+	if len(errs) > 0 {
+		return fmt.Errorf("failed while closing threads; err(s): %q", errs)
+	}
+
+	t.cancel()
+	return nil
+}
+
+func (t *threads) Host() host.Host {
+	return t.host
+}
+
+func (t *threads) DAGService() format.DAGService {
+	return t.dagService
+}
+
+func (t *threads) Add(ctx context.Context, body format.Node, opts ...tserv.AddOption) (l peer.ID, r thread.Record, err error) {
+	// Get or create a log for the new node
+	settings := tserv.AddOptions(opts...)
+	lg, err := t.getOrCreateOwnLog(settings.Thread)
+	if err != nil {
+		return
+	}
+
+	// Write a node locally
+	coded, err := t.createNode(ctx, body, lg, settings)
+	if err != nil {
+		return
+	}
+
+	// Push out the new record
+	err = t.service.push(ctx, coded, settings.Thread, lg.ID, settings)
+	if err != nil {
+		return
+	}
+
+	return lg.ID, coded, nil
+}
+
+func (t *threads) Put(ctx context.Context, rec thread.Record, opts ...tserv.PutOption) error {
```

Comment:
Top-level API for putting existing records into a log. I think this will be reworked into an internal method.

Reply:
A Record contains the Event? If the Event contains the sequence info, how will it rework the Event's sequencing method (i.e. the clock record) for the target Log here? Am I reading it right below that it just pulls the core bits of the Event (header/body) and wraps them in a new event for the target log?

Reply:
Yep, events are within the record, which is the top level node. Clocks will have to be handled at the model +/ event store level. Here we just put the record, which falls into whatever place is described by its prev link.